docs(framework) Add how-to guide for using Flower `ClientApp`

chongshenng opened the pull request:
<!--
Thank you for opening a pull request (PR)!

Contribution guidelines: https://github.com/adap/flower/blob/main/CONTRIBUTING.md
-->

## Issue

### Description
Our Flower deployment-readiness lack official feature documentation. 
<!--
Describe the problem addressed by this PR.

Example: The variable name `rnd` could be misinterpreted as an abbreviation of *random*, but it refers to the current server round.
-->

### Related issues/PRs
N/A
<!--
Link issues and/or PRs that are related to this PR.

Example: Fixes #123. See also #456 and #789.
-->

## Proposal

### Explanation
Add a how-to guide highlighting the deployment-readiness feature of Flower.
<!--
Explain the changes and how they improve the issue described above.

Example: The variable `rnd` was renamed to `server_round` to improve readability.
-->

### Checklist

- [X] Implement proposed change
- ~~[ ] Write tests~~
- [X] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [X] Update the changelog entry below
- [x] Make CI checks pass

<!--
Inside the following 'Changelog entry' section, you should put the description of your changes that will be added to the changelog alongside your PR title.

If the section is completely empty (without any token) or non-existant, the changelog will just contain the title of the PR for the changelog entry, without any description. If the section contains some text other than tokens, it will use it to add a description to the change. If the section contains one of the following tokens it will ignore any other text and put the PR under the corresponding section of the changelog:

<general> is for classifying a PR as a general improvement.
<skip> is to not add the PR to the changelog
<baselines> is to add a general baselines change to the PR
<examples> is to add a general examples change to the PR
<sdk> is to add a general sdk change to the PR
<simulations> is to add a general simulations change to the PR

Note that only one token should be used.
-->

### Changelog entry

<skip>

### Any other comments?
N/A
<!--
Please be aware that it may take some time until the maintainers can review the PR.
Smaller PRs with good descriptions can be considered much more easily.

If you have an urgent request or question, please use the Flower Slack:

    https://flower.ai/join-slack/ (channel: #contributions)

Thank you for contributing to Flower!
-->







<!--
Thank you for opening a pull request (PR)!

Please rename your PRs following this [format](https://flower.ai/docs/framework/contributor-tutorial-contribute-on-github.html#pr-title-format).
Contribution guidelines: https://github.com/adap/flower/blob/main/CONTRIBUTING.md
-->

## Issue

Our Flower `ClientApp` API lack official feature documentation. 

### Description

<!--
Describe the problem addressed by this PR.

Example: The variable name `rnd` could be misinterpreted as an abbreviation of *random*, but it refers to the current server round.
-->

### Related issues/PRs

<!--
Link issues and/or PRs that are related to this PR.

Example: Fixes #123. See also #456 and #789.
-->

## Proposal

### Explanation

<!--
Explain the changes and how they improve the issue described above.

Example: The variable `rnd` was renamed to `server_round` to improve readability.
-->

### Checklist

- [X] Implement proposed change
- ~~[ ] Write tests~~
- [X] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Any other comments?

<!--
Please be aware that it may take some time until the maintainers can review the PR.
Smaller PRs with good descriptions can be considered much more easily.

If you have an urgent request or question, please use the Flower Slack:

    https://flower.ai/join-slack/ (channel: #contributions)

Thank you for contributing to Flower!
-->

mohammadnaseri commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,304 @@
+* :code:`app.evaluate()`
+* :code:`app.query()`
+
+which, you guessed it, runs the training, evaluation, and query on client-side data. 
```
I would suggest to add some explanation about these. specially the query which is not explained at all..

chongshenng replied:
@mohammadnaseri Do you mean additional explainer text next to these bullet points? In each of the sections for `app.train()` and `app.evaluate()`, there are explanations for the methods, which I think are sufficient.

I've prioritized `app.train()` and `app.evaluate()` based on a discussion I had with @danieljanes.

mohammadnaseri replied:
I meant mainly for `query` as it is mentioned but not explained. maybe we can remove it until we have some explanation for it.

mohammadnaseri commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,304 @@
+
+Now that we’ve implemented client training, let’s walk through how to register an evaluation function in the :code:`ClientApp`.
+
+The structure of the evaluate function is the same as as :code:`app.train()`:
```
I suggest also mentioning the decorator `@app.evaluate()`

chongshenng commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,304 @@
+Use :code:`ClientApp`
+=====================
+
+With Flower 1.8, we introduced the :code:`ClientApp` and :code:`ServerApp`. These are new features in Flower that allow more flexibility for you to run your federated learning projects, such as adding intermediate operations in your workflows :doc:`using Mods <how-to-use-built-in-mods>`, maximizing compute usage by :doc:`running multiple projects concurrently <how-to-use-multirun>`, and speeding  ...
```
@mohammadnaseri Indeed! Those are links to other `*.md` pages in the project using the [`doc` role](https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html#the-doc-role). The multirun and deployment readiness pages are under review in #3393 and #3395 and will have those page names when merged :).

chongshenng commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,304 @@
+What is a :code:`ClientApp`?
+----------------------------
+
+The |clientapp_link|_ is an application that can run various tasks on clients, such as federated learning. It is launched on demand by the :code:`SuperNode`. The communication pattern between the :code:`SuperNode` and :code:`ClientApp` uses :code:`Message` and :code:`Context` objects containing all the information needed to run the tasks. Don’t worry about the details of the :code:`Message` and : ...
```
> what is the underline after |clientapp_link| for? 

This is a nifty trick I found [here](https://github.com/jgm/pandoc/issues/3973#issuecomment-337087394) that generates link text as literals. See how `ClientApp()` is rendered in our [upgrade to Flower Next page](https://flower.ai/docs/framework/how-to-upgrade-to-flower-next.html#required-changes). 

> why do you mean by "various tasks on clients, such as federated learning"? like federated learning is a task?

Yes, that's what I meant. I agree that FL takes precedence, but as far as I understand, there is a plan to include federated analytics. 

>  But just to be consistent else where as well;

Sure, I can replace that.

mohammadnaseri replied:
I would say `model training` instead of `federated learning`, but that is just my opinion.

chongshenng commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,304 @@
+    # file: client.py
+
+    class CifarClient(NumPyClient):
+        def get_parameters(self, config):
```
@mohammadnaseri That makes sense, and that's what I wrote for the low-level API section. The `get_params` and `set_params` are for the compatibility mode where there are existing `Client`s or `NumPyClient`s. Does that make sense?

mohammadnaseri replied:
Ah. I see. thanks for clarifying.

chongshenng commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,304 @@
+
+    # file: client.py
+
+    def client_fn() -> Client:
```
Hmm, indeed we do. Though as far as I understand, `cid` is only used in `run_simulation`. I'll add it for now.

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+Using high-level API
+--------------------
+
+This approach is the compatibility mode. It allows us to reuse any client-code that has been implemented using :code:`NumpyClient` or :code:`Client` classes with a few additional lines of code. Let’s take the :doc:`PyTorch Quickstart <tutorial-quickstart-pytorch>` code for example. In that example, we have included a :code:`CifarClient` class that inherits from the :code:`NumpyClient` convenience ...
```
I'm wondering if it would be worth linking to the "migration guide" page you added recently?  maybe at the bottom of this section. e.g. "please refer to <migration guide> to see how to migrate nos just ClientApp but also ServerApp and your Simulation". Would something like this make sense? (maybe as a admotion note or similar)

chongshenng replied:
Good idea, I've added an admonition at the start of this section.

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+Using low-level API
+-------------------
+
+With Flower 1.8, we provide a set of low-level APIs to allow more versatile ways to implement any functionality that we like in the :code:`ClientApp`. The three functions that can be registered in a :code:`ClientApp` are:
```
You say "three functions" but below you list only two of them. `app.query()` is missing

chongshenng replied:
Good catch, I've fixed it. The plan is to update the guide later with `app.query()` - see [this other thread](https://github.com/adap/flower/pull/3510#discussion_r1616241795).

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+.. admonition:: Important
+    :class: important
+
+    As we continuously enhance Flower at a rapid pace, we'll periodically update the functionality and this how-to document. Please feel free to share any feedback with us!
```
Maybe worth adding link to slack at the end of this sentence? or something?

chongshenng replied:
Added the link and moved both the admonition and link to the end of the guide.

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+
+    As we continuously enhance Flower at a rapid pace, we'll periodically update the functionality and this how-to document. Please feel free to share any feedback with us!
+
+Next, we have a key part in the low-level API, which is the message passing format. In Flower 1.8, we introduced |message_link|_ and |context_link|_ objects. These are abstractions that unify the message formats that are relayed back and forth in a Flower project. When compared to the high-level example above, objects such as :code:`config` and :code:`parameter` need to be separately configured a ...
```
A coupe of suggestions:
- "[messages, context] are abstractions to unify the message formats" --> i think it would be confusing to say the context is part of messaging format. This is because the context is local to the `SuperNode`, never leaves it. I think of it as some form of "in memory state" that persists until the`SuperNode` is shutdown. (in the case of simulations it persists until the simulation is over)
- "[...] which can make the workflow un-intuitive." -> i think the main limitation is that the workflow is a bit rigid. maybe we can adjust the wording ? "intuitive" is a bit relative 😅  (to me is very intuitive).

chongshenng replied:
@jafermarq 
* > This is because the context is local to the SuperNode

  Thanks for clarifying, I had the `LegacyContext` example in mind when I wrote this. I'll reword to clarify that context is the local state of the `SuperNode.
*   Sure, I'll reword too 😄

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+.. admonition:: Tip
+    :class: note
+
+    The Message and Context objects contain data such as parameters, metrics, configs, and identifiers of the current run. All of these information is used by the :code:`ClientApp` to determine what and how to run the task.
```
I think it's worth highlight that the `Context` doesn't store anything (i.e. it's empty) unless the user, via the `ClientApp` or a `mod`, decides to save something to it

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+In short:
+
+#. :code:`Message` is a simplified messaging pattern between :code:`ClientApp` and :code:`ServerApp`, and is consumed within a round.
+#. :code:`Context` persists messages across rounds.
```
`Context` is a `RecordSet` object, which is the same type of object carrying the bulk of the information inside a `Message` (i.e. `Message.content` is of type `RecordSet`) -- not sure if this level of detail is needed here but I feel it's better not to define `Context` with the word "message".

chongshenng replied:
Good point, I've also associated `Context` with information in the sentence above. Have revised the wordings.

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+:code:`app.train()`
+~~~~~~~~~~~~~~~~~~~
+
+The :code:`app.train()` decorator registers an end-to-end training function on a client. The following steps outline how to implement it. First, import the necessary modules and register the function :code:`train()` using the decorator :code:`@app.train()`:
```
What does "end-to-end" mean here?

chongshenng replied:
I meant from the start to the end of a single FL round of training. I'll clarify it to say it "registers a single training task on a client, e.g. for FL, it executes one FL round of model training on a client".

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+    # load local training and validation data
+    train_loader, val_loader, _ = load_data()
+
+For simplicity, we have omitted the implementations for :code:`Net()` and :code:`load_data()` , but you can refer to the :code:`quickstart-pytorch` code for similar implementations.
```
```suggestion
For simplicity, we have omitted the implementations for :code:`Net()` and :code:`load_data()`, but you can refer to the :code:`quickstart-pytorch` code for similar implementations.
```

jafermarq commented on doc/source/how-to-use-clientapp.rst:
```diff
@@ -0,0 +1,303 @@
+
+That’s it! You now have a working Flower :code:`ClientApp` that initializes the model, loads local data, trains the model, and returns the updated model and metrics to the :code:`ServerApp`. Note that the training workflow is dramatically simplified and can be intuitively implemented end-to-end.
+
+For completeness, you can implement the utility functions referenced in the code snippets above as follows:
```
Worth stating that soon we'll be providing these utility functions so it becomes trivial to convert a Pythorch/TF/whatever model to a `ParametersRecord` and back.

jafermarq reviewed (changes requested):
@chongshenng this doc is gold! I left a few comments. Happy to clarify or sync about these. I'd add addition to the `@app.train` and `@app.evaluate` code lines below: instantiate the model IIF it's `state_dict` is not found in the context (this means that the first round you'll likely save the model to the state) Since the `Context` is of type `RecordSet` you can easily reuse the utility functions you have to send the model parameters.

chongshenng commented:
@jafermarq I've addressed your comments and added the usage of `Context.content.metrics_record` to both `@app.train()` an `@app.evaluate()`. Let me know if there're anything else!
